### DISCUSSION TITLE
Talking to Postgres Through Java 16 Unix-Domain Socket Channels - Gunnar Morling

### DISCUSSION THREAD

**Discussion by giscus** [gunnarmorling/discussions.morling.dev#287] [Announcements]
# Talking to Postgres Through Java 16 Unix-Domain Socket Channels - Gunnar Morling

Update Feb 5: This post is discussed on Hacker News

Reading a blog post about what’s coming up in JDK 16 recently,
I learned that one of the new features is support for Unix domain sockets (JEP 380).
Before Java 16, you’d have to resort to 3rd party libraries like jnr-unixsocket in order to use them.
If you haven’t heard about Unix domain sockets before,
they are "data communications [endpoints] for exchanging data between processes executing on the same host operating system".
Don’t be put off by the name btw.;
Unix domain sockets are also supported by macOS and even Windows since version 10.

https://www.morling.dev/blog/talking-to-postgres-through-java-16-unix-domain-socket-channels/

**Comment by shrik-dr** [comment]
Any idea whether we can use the junk sockets with  org.apache.commons.dbcp.BasicDataSourceFactory ??
I tried setting the 'socketFactory' parameter in properties and created the DataSource. But not able to get the connection to DB

**Comment by jedvardsson** [comment]
The reason for why your attempt with SocketChannel.socket() is not working is that the socket returned does not support Unix domain sockets. There is an explicit check to only allow INET sockets.
```
    @Override
    public Socket socket() {
        synchronized (stateLock) {
            if (socket == null) {
                if (isNetSocket()) {
                    socket = SocketAdaptor.create(this);
                } else {
                    throw new UnsupportedOperationException("Not supported");
                }
            }
            return socket;
        }
    }
```

However you can create your own SocketAdaptor class which supports Unix sockets. The problem is that I think Socket was developed with TCP in mind only. Unix sockets are not tcp. So there need to be some modification to the socket methods that are TCP related. Below is some code that does the job. 

https://gist.github.com/jedvardsson/7ba7bbc94b4951f82da4b590ace725d2
